### ISSUE TITLE
Additional repository support ideas

### ISSUE THREAD

**Issue by AMDmi3** [repology/repology-updater#1]
Please share any ideas on what additional repositories we can support. A description on how to fetch all package data from specific repository is preferred. Approved repositories with determined fetching algorithm are split to separate bugs and eventually implemented.
### Classic *nix package repositories
- Any RPM repos
  - :white_check_mark: ~~Fedora~~ (see #36)
  - :white_check_mark: ~~OpenSUSE~~ (see #44)
  - :white_check_mark: ~~AltLinux Sisyphus~~ (see #24)
  - :white_check_mark: ~~Fedora EPEL~~
- :white_check_mark: ~~slackware~~ (#331)
- :white_check_mark: ~~[homebrew](http://brew.sh)~~ (see #198)
- :white_check_mark: ~~DragonFlyBSD's dports (pretty much the same as FreeBSD ports minus some packages which don't build on DragonFly)~~
- :bulb: [VectorLinux](http://vectorlinux.com/)

From Fedora [release-monitoring](https://release-monitoring.org/distros/)
- :white_check_mark: ~~[Alpine linux](https://pkgs.alpinelinux.org/packages)~~
- :white_check_mark: ~~[CRUX](https://crux.nu/portdb/) whatever that is~~

Unsorted
- :white_check_mark: ~~[SliTaz](http://hg.slitaz.org/wok/)~~ (#664)
- :bulb: [NuTyX](https://www.nutyx.org/en/) (only shell based pkgbuilds are available)
- :x: ~~[Sabotage](https://github.com/sabotage-linux/sabotage/)~~ (do not use versions)
- :white_check_mark: ~~[salix](http://www.salixos.org/)~~
- :white_check_mark: ~~[solus](https://solus-project.com/) (https://git.solus-project.com/, https://packages.solus-project.com/unstable/)~~ (#341)

### Other platforms
- :white_check_mark: ~~[NixOS](https://github.com/NixOS/nixpkgs)~~
- :white_check_mark: ~~[YACP](https://github.com/fd00/yacp) (Yet another cygwin ports)~~ (though project is somewhat inactive)
- :white_check_mark: ~~Rosa~~
- :white_check_mark: ~~[GuixSD](https://www.gnu.org/software/guix/packages)~~
- :bulb: [OpenPandora](https://repo.openpandora.org/?page=all&s=name&ipp=all)
- :white_check_mark: ~~[buckaroo](https://github.com/LoopPerfect/buckaroo-recipes)~~ :+1: json recipes, :-1: custom naming (fixed by some rules). Also looks dead already (true, 75% outdatedness).
- :white_check_mark: ~~MSYS2~~ (see #262 )

Since these will contain too many unrelevant unique packages, doable as shadow repos:
- :white_check_mark: ~~Chocolatey~~ (see #43)
- :white_check_mark: ~~[F-Droid](https://f-droid.org/repo/index.xml)~~ (only a handful of packages manually whitelisted, too much android garbage)
### Upstream repos
 
Doable as shadow repos as well
- :white_check_mark: ~~CPAN (perl packages)~~
- :white_check_mark: ~~PyPi (python packages)~~
- :white_check_mark: ~~RubyGems~~
- :bulb: Others? (node, etc).
- :bulb: GitHub, for projects which fetch from there. Need feedback loop here (parse normal repos -> get github urls -> parse github urls for latest tags)
### New version detectors
- :bulb: FreeBSD's [portscout](http://portscout.freebsd.org/)
- :bulb: [pkgsrc's thingy](http://www.ki.nu/~makoto/pkgsrc/check-update/20160928-17/00_whole.html)
- :bulb: [AllMyChanges.com](https://allmychanges.com/) just for completeness, usefullness is questionable: it's focused on iOS apps and stuff which intersects with *nix software is fairly outdated

...more ideas?

**Comment by AMDmi3** [comment]
Homebrew doesn't seem to specify version numbers for formulas. Parse it from url?

**Comment by jbicha** [comment]
Ubuntu

**Comment by AMDmi3** [comment]
> Ubuntu

It is already supported (repo is commented out though). It'll be more easily available from command line with the next update I'm going to push quite soon. I didn't plan to support it on the website though, as it seemed to have the same versions as Debian, but I see now that's not the case, so I'll add it.

**Comment by jktrigg** [comment]
SuSE

**Comment by AMDmi3** [comment]
> SuSE

Any ideas how to get information on its packages?

**Comment by jktrigg** [comment]
Tumbleweed (equivalent to CURRENT): http://download.opensuse.org/tumbleweed/repo/oss/suse/{noarch,x86_64}
Leap (equivalent to STABLE): http://download.opensuse.org/distribution/leap/42.1/repo/oss/suse/{noarch,x86_64}

**Comment by AMDmi3** [comment]
These contain to little info. Well, they are parsable so can try, but I can only parse package names vesions out of it, while the project already uses maintainer field and will use more fields in future. Isn't there a way to get all .spec files?

**Comment by AMDmi3** [comment]
@jktrigg I've added OpenSUSE support they way you've described. It's not really suitable for comparison though, as these binary packages have different naming schema: for example, for each ```libfoo``` in other repos we'll have ```libfooN``` ```libfoo-devel``` and ```libfooN-32bit``` here.

**Comment by AMDmi3** [comment]
@jbicha FYI, Ubuntu is fully supported
@jktrigg FYI, OpenSUSE is fully supported

**Comment by vadi2** [comment]
I'd like to see homebrew supported - I see http://braumeister.org is able to show the version information.

**Comment by AMDmi3** [comment]
@vadi2 me too. Homebrew author is defiantly unsupportive, so yes, braumeister.org is our hope. I've already filed https://github.com/koraktor/braumeister.org/issues/45, you could :+1: it

**Comment by AMDmi3** [comment]
Homebrew support is now tracked in separate issue #198

**Comment by dg0yt** [comment]
How about MSYS2? It is based on Arch Linux' pacman.

**Comment by AMDmi3** [comment]
Are there any links?

**Comment by dg0yt** [comment]
Apart from http://www.msys2.org/, there is the repository at http://repo.msys2.org/, which has index files such as http://repo.msys2.org/msys/x86_64/msys.files.tar.gz. But I'm not an expert for Arch or pacman.

**Comment by AMDmi3** [comment]
Nice! MSYS2 support added.

**Comment by dg0yt** [comment]
This was quick!

However, I start to understand that the [MSYS2 "subsystems"](https://github.com/msys2/msys2/wiki/MSYS2-introduction#msys2-susbsystems) need to be treated individually:

> The mingw subsystems provide native Windows programs and are the main focus of the project. 

> The msys2 subsystem provides an emulated mostly-POSIX-compliant environment for building software, package management, and shell scripting.

Most userland software resides in the mingw subsystem. For example, [repology doesn't indicate](https://repology.org/metapackage/gdal/versions) that there is a [gdal package in the mingw subsystem](https://github.com/Alexpux/MINGW-packages/tree/master/mingw-w64-gdal) at the moment. So probably there needs to be 
- MSYS2 mingw
- MSYS2 msys2 (as is now)

(Note: For mingw, there is another minor difference, mingw32 vs. mingw64, but I think this can be neglected.)

**Comment by AMDmi3** [comment]
On the second glance, I don't think we can support this yet. We need to parse PKGBUILDs from git repos, and that depends on #166. With processed databases from repo.msys2.org it's impossible to filter out subpackages properly.

**Comment by AMDmi3** [comment]
Let's move MSYS2 discussion to #262

**Comment by tkelman** [comment]
Cygwin: http://mirrors.mit.edu/cygwin/x86_64/setup.ini, http://mirrors.mit.edu/cygwin/x86/setup.ini

**Comment by AMDmi3** [comment]
> Cygwin

Let's discuss it in #269

**Comment by ikeydoherty** [comment]
I notice you've got Solus listed there. Note our repos are over at https://dev.solus-project.com

If you need us to provide some parsers or an additional indexing method, let us know. Long story short for the published index in unstable the Source is shared among a package group, and History.Updates[0].Version is the published version.

**Comment by AMDmi3** [comment]
@ikeydoherty I think I've already investigated Solus index. What stopped me from adding support for it right away is the fact that the index lists binary packages, while repology prefers source packages. While it's possible to work with binary packages, merging (libfoo, libfoo-devel, libfoo-dbginfo, libfoo-docs, libfoo-32bit-*) etc. requires extra work and is error prone. It could be also possible to get package names from Source.Name, but I haven't investigated yet if that would work. Parsing package.yml files from sources could work too, but I see no way to download them in one go yet. Downloading a repository for each package is not feasible.

**Comment by ikeydoherty** [comment]
Yep, I'd agree with you on that one. After I'm back from NYC I'll be finishing off the other parts of the new Solus build infra. One of those aspects will require us to have a server performing pulls/validation on all of our source repositories so that we can enable interdependency analysis in build queueing. Honestly I'd planned to defer it for a bit, but this might give the the excuse I need. With that mass-source-import running on our servers, we could simple emit a singular source index for the totality of our repos.

If you're interested in that lemme know, and what sort of formats you'd like to be using (I'm pretty sure we all want XML feeds to die at this point.)

**Comment by AMDmi3** [comment]
I'm interested in supporting as many repos as possible :) If there's also collaboration from the repository side, that's absolutely great. JSON dump would be most convenient for me.

**Comment by AMDmi3** [comment]
Let's continue discussion in #341

**Comment by davidak** [comment]
### elementary OS 0.4 Loki

based on Ubuntu (deb based)

```
deb http://ppa.launchpad.net/elementary-os/stable/ubuntu xenial main
deb http://ppa.launchpad.net/elementary-os/os-patches/ubuntu xenial main
```

They have some original software for their distro, also packaged in AUR and Fedora. Also they will have a new release probably soon, but it will be great to see the progress.

**Comment by jbicha** [comment]
I don't know if elementary OS has diverged enough for it to be useful.

Here is a different view of the packages for their 0.4 release:

[stable](https://launchpad.net/~elementary-os/+archive/ubuntu/stable/+packages?field.series_filter=xenial)

[os-patches](https://launchpad.net/~elementary-os/+archive/ubuntu/os-patches/+packages?field.series_filter=xenial)

**Comment by jbicha** [comment]
Well on second thought, there are other similarly small derivative distros listed so never mind.

**Comment by AMDmi3** [comment]
Elementary doesn't look usable as it uses custom mangled version scheme (appends rXXX).

**Comment by AMDmi3** [comment]
Locking this. Please submit new ideas as separate issues.
